core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java

```diff
@@ -0,0 +1,267 @@
+
+  private final ConcurrentMap<KeyspaceTableNamePair, NavigableSet<Tablet>> mapping;
+
+  public DefaultTabletMap(ConcurrentMap<KeyspaceTableNamePair, NavigableSet<Tablet>> mapping) {
+    this.mapping = mapping;
+  }
+
+  public static DefaultTabletMap emptyMap() {
+    return new DefaultTabletMap(new ConcurrentHashMap<>());
+  }
+
+  @Override
+  public Map<KeyspaceTableNamePair, NavigableSet<Tablet>> getMapping() {
+    return mapping;
+  }
+
+  @Override
+  public Set<UUID> getReplicas(String keyspace, String table, long token) {
+    KeyspaceTableNamePair key = new KeyspaceTableNamePair(keyspace, table);
+
+    if (mapping == null) {
+      logger.trace("This tablets map is null. Returning empty set.");
+      return Collections.emptySet();
+    }
+
+    NavigableSet<Tablet> set = mapping.get(key);
+    if (set == null) {
+      logger.trace(
+          "There is no tablets for {}.{} in this mapping. Returning empty set.", keyspace, table);
+      return Collections.emptySet();
+    }
+    Tablet row = mapping.get(key).ceiling(DefaultTablet.malformedTablet(token));
+    if (row == null || row.getFirstToken() >= token) {
+      logger.trace(
+          "Could not find tablet for {}.{} that owns token {}. Returning empty set.",
+          keyspace,
+          table,
+          token);
+      return Collections.emptySet();
+    }
+
+    HashSet<UUID> result = new HashSet<>();
+    for (HostShardPair hostShardPair : row.getReplicas()) {
+      result.add(hostShardPair.getHost());
+    }
+    return result;
+  }
+
+  /**
+   * Updates tablet map with parsed tablets-routing-v1 custom payload. Expects the tuple's
+   * underlying structure to correspond to {@code TupleType(LongType, LongType,
+   * ListType(TupleType(UUIDType, Int32Type)))}. Handles removing outdated tables that intersect
+   * with the one about to be added.
+   *
+   * @param keyspace the keyspace of the table
+   * @param table the table name
+   * @param tupleValue the parsed payload
+   */
+  @Override
+  public void addTablet(String keyspace, String table, TupleValue tupleValue) {
+    KeyspaceTableNamePair ktPair = new KeyspaceTableNamePair(keyspace, table);
+
+    long firstToken = tupleValue.getLong(0);
+    long lastToken = tupleValue.getLong(1);
+
+    logger.trace(
+        "Received tablets routing V1 payload: {}.{} range {}-{}",
+        keyspace,
+        table,
+        firstToken,
+        lastToken);
+
+    Set<HostShardPair> replicas = new HashSet<>();
+    List<TupleValue> list = tupleValue.getList(2, TupleValue.class);
+    assert list != null;
+    for (TupleValue tuple : list) {
+      HostShardPair hostShardPair = new HostShardPair(tuple.getUuid(0), tuple.getInt(1));
+      replicas.add(hostShardPair);
+    }
+
+    // Working on a copy to avoid concurrent modification of the same set
+    NavigableSet<Tablet> existingTablets =
+        new TreeSet<>(mapping.computeIfAbsent(ktPair, k -> new TreeSet<>()));
```

Comment:
Wait wait wait wait..... we need to copy the whole tablet list for the table (which afaik may contain up to 100k tablets) for each tablet insertion? So inserting N tablet's has the complexity O(N^2)? With N being 10^5 this is 10^10 operations. This is not acceptable

Reply:
10^5 indeed does sound like a lot. Removed the copying, since it seems unnecessary in current solution anyway.